File: app/src/main/java/com/alvarlagerlof/koda/Login/ForgotPasswordActivity.java
package com.alvarlagerlof.koda.Login;

import android.app.Activity;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.design.widget.TextInputEditText;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.view.animation.DecelerateInterpolator;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.alvarlagerlof.koda.R;
import com.alvarlagerlof.koda.RemoteConfigValues;
import com.bumptech.glide.Glide;
import com.google.firebase.crash.FirebaseCrash;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import okhttp3.FormBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

/**
 * Created by alvar on 2016-07-02.
 */
public class ForgotPasswordActivity extends AppCompatActivity {

    TextInputEditText email;
    Button nextButton;
    ProgressBar progressBar;


    LinearLayout emailView;
    LinearLayout resultView;
    LinearLayout errorView;

    TextView resultTextView;
    TextView errorText;

    Toolbar toolbar;


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.login_forgot_pass_activity);

        email = (TextInputEditText) findViewById(R.id.email);
        nextButton = (Button) findViewById(R.id.next_button);
        progressBar = (ProgressBar) findViewById(R.id.progress_bar);


        emailView = (LinearLayout) findViewById(R.id.email_view);
        resultView = (LinearLayout) findViewById(R.id.result_view);
        errorView = (LinearLayout) findViewById(R.id.error_view);


        resultTextView = (TextView) findViewById(R.id.result_text);
        errorText = (TextView) findViewById(R.id.error);

        toolbar = (Toolbar) findViewById(R.id.toolbar);

        setSupportActionBar(toolbar);
        toolbar.setNavigationIcon(R.drawable.ic_close);



        Glide.with(this)
                .load(RemoteConfigValues.URL_LOGIN_FORGOT_IMAGE)
                .into((ImageView) findViewById(R.id.background));

        findViewById(R.id.background).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                InputMethodManager inputMethodManager = (InputMethodManager) ForgotPasswordActivity.this.getSystemService(Activity.INPUT_METHOD_SERVICE);
                inputMethodManager.hideSoftInputFromWindow(findViewById(R.id.background).getWindowToken(), 0);
            }
        });


    }



    public void next(View view) {
        progressBar.setVisibility(View.VISIBLE);
        new PasswordAsync().execute();

    }

    public void finish(View view) {
        InputMethodManager inputMethodManager = (InputMethodManager) ForgotPasswordActivity.this.getSystemService(Activity.INPUT_METHOD_SERVICE);
        inputMethodManager.hideSoftInputFromWindow(findViewById(R.id.background).getWindowToken(), 0);
        finish();
    }


    class PasswordAsync extends AsyncTask<Void, Integer, String> {

        String emailString;

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            emailString = email.getText().toString();
        }


        protected String doInBackground(Void... arg0) {

            OkHttpClient client = new OkHttpClient.Builder().build();

            RequestBody formBody = new FormBody.Builder()
                    .add("email", emailString)
                    .add("verification_reset", "7")
                    .add("headless", "thisIsSet")
                    .build();


            Request request = new Request.Builder()
                    .url(RemoteConfigValues.URL_LOGIN_FORGOT)
                    .post(formBody)
                    .build();

            String result = null;

            try {
                Response response = client.newCall(request).execute();
                result = response.body().string();
                response.body().close();
            } catch (IOException e) {
                e.printStackTrace();
            }


            return result;
        }

        @Override
        protected void onPostExecute(String s) {
            super.onPostExecute(s);

            if (s != null) {
                try {

                    JSONObject jsonObject = new JSONObject(s);

                    if (jsonObject.getString("success").equals("true")) {

                        resultTextView.setText(jsonObject.getString("message"));

                        Animation fadeOut = new AlphaAnimation(1, 0);
                        fadeOut.setInterpolator(new AccelerateInterpolator()); //and this
                        fadeOut.setDuration(500);
                        fadeOut.setAnimationListener(new Animation.AnimationListener() {
                            @Override
                            public void onAnimationStart(Animation animation) {

                            }

                            @Override
                            public void onAnimationEnd(Animation animation) {
                                emailView.setVisibility(View.INVISIBLE);
                            }

                            @Override
                            public void onAnimationRepeat(Animation animation) {

                            }
                        });


                        Animation fadeIn = new AlphaAnimation(0, 1);
                        fadeIn.setInterpolator(new DecelerateInterpolator()); //projects_add this
                        fadeIn.setDuration(500);
                        fadeIn.setStartOffset(1000);
                        fadeIn.setFillAfter(true);


                        resultTextView.setVisibility(View.VISIBLE);

                        emailView.startAnimation(fadeOut);
                        resultView.startAnimation(fadeIn);


                    } else {
                        errorText.setText(jsonObject.getString("message"));

                        Animation fadeOut = new AlphaAnimation(1, 0);
                        fadeOut.setInterpolator(new AccelerateInterpolator());
                        fadeOut.setDuration(500);
                        fadeOut.setAnimationListener(new Animation.AnimationListener() {
                            @Override
                            public void onAnimationStart(Animation animation) {
                            }

                            @Override
                            public void onAnimationEnd(Animation animation) {
                                emailView.setVisibility(View.GONE);
                                resultView.setVisibility(View.GONE);
                            }

                            @Override
                            public void onAnimationRepeat(Animation animation) {
                            }
                        });


                        emailView.startAnimation(fadeOut);


                        Animation fadeIn = new AlphaAnimation(0, 1);
                        fadeIn.setInterpolator(new DecelerateInterpolator()); //projects_add this
                        fadeIn.setDuration(500);
                        fadeIn.setStartOffset(500);
                        fadeIn.setFillAfter(true);

                        errorView.setVisibility(View.VISIBLE);
                        errorView.startAnimation(fadeIn);
                    }

                } catch (JSONException e) {
                    e.printStackTrace();
                    FirebaseCrash.report(e);
                }

            }
        }
    }


    public void reset(View view) {

        final Animation fadeIn = new AlphaAnimation(0, 1);
        fadeIn.setInterpolator(new DecelerateInterpolator()); //projects_add this
        fadeIn.setDuration(500);
        fadeIn.setStartOffset(500);
        fadeIn.setFillAfter(true);

        Animation fadeOut = new AlphaAnimation(1, 0);
        fadeOut.setInterpolator(new AccelerateInterpolator());
        fadeOut.setDuration(500);
        fadeOut.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {}

            @Override
            public void onAnimationEnd(Animation animation) {
                emailView.setVisibility(View.VISIBLE);
                progressBar.setVisibility(View.INVISIBLE);
                emailView.startAnimation(fadeIn);
                errorView.setVisibility(View.GONE);
            }

            @Override
            public void onAnimationRepeat(Animation animation) {}
        });


        errorView.startAnimation(fadeOut);

    }


    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle item selection
        switch (item.getItemId()) {
            case android.R.id.home:
                InputMethodManager inputMethodManager = (InputMethodManager) ForgotPasswordActivity.this.getSystemService(Activity.INPUT_METHOD_SERVICE);
                inputMethodManager.hideSoftInputFromWindow(findViewById(R.id.background).getWindowToken(), 0);
                finish();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

}

File: app/src/main/java/com/alvarlagerlof/koda/Login/ForgotPasswordActivity.kt
package com.alvarlagerlof.koda.Login

import android.content.Intent
import android.os.AsyncTask
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem
import android.view.View
import android.view.animation.AccelerateInterpolator
import android.view.animation.AlphaAnimation
import android.view.animation.Animation
import android.view.animation.DecelerateInterpolator
import com.alvarlagerlof.koda.Extensions.hideKeyboard
import com.alvarlagerlof.koda.Extensions.isConnected
import com.alvarlagerlof.koda.Extensions.showKeyboard
import com.alvarlagerlof.koda.R
import com.alvarlagerlof.koda.Vars
import com.bumptech.glide.Glide
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.crash.FirebaseCrash
import kotlinx.android.synthetic.main.login_forgot_pass_activity.*
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException

/**
 * Created by alvar on 2016-07-02.
 */
class ForgotPasswordActivity : AppCompatActivity() {

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.login_forgot_pass_activity)

        setSupportActionBar(toolbar)
        toolbar.setNavigationIcon(R.drawable.ic_close_white)


        Glide.with(this)
                .load(Vars.URL_LOGIN_FORGOT_IMAGE)
                .into(background)

        background.setOnClickListener {
            background.hideKeyboard()
        }

        FirebaseAnalytics.getInstance(this).logEvent("login_forgot_open", Bundle())


    }


    @Suppress("UNUSED_PARAMETER")
    fun next(view: View) {
        email.hideKeyboard()
        if (this.isConnected()) {
            pacman_progress.visibility = View.VISIBLE
            PasswordAsync().execute()
        } else {
            AlertDialog.Builder(this)
                    .setTitle("Ingen ansluting")
                    .setMessage("Gå in i inställingar och se till att du har Wifi eller mobildata på")
                    .setPositiveButton("Ok") { dialog, _ -> dialog.cancel() }
                    .show()
        }



    }

    @Suppress("UNUSED_PARAMETER")
    fun finish(view: View) {
        background.hideKeyboard()
        finish()
    }


    internal inner class PasswordAsync : AsyncTask<Void, Int, String>() {

        var emailString: String = ""

        override fun onPreExecute() {
            super.onPreExecute()
            emailString = email.text.toString()
        }

        override fun doInBackground(vararg arg0: Void): String {

            val client = OkHttpClient.Builder().build()

            val formBody = FormBody.Builder()
                    .add("email", emailString)
                    .add("verification_reset", "7")
                    .add("headless", "thisIsSet")
                    .build()


            val request = Request.Builder()
                    .url(Vars.URL_LOGIN_FORGOT)
                    .post(formBody)
                    .build()

            var result: String? = null

            try {
                val response = client.newCall(request).execute()
                result = response.body()!!.string()
                response.body()!!.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }


            return result as String
        }

        override fun onPostExecute(s: String?) {
            super.onPostExecute(s)

            if (s != null) {
                try {

                    val jsonObject = JSONObject(s)

                    if (jsonObject.getString("success") == "true") {

                        FirebaseAnalytics.getInstance(applicationContext).logEvent("login_forgot_successful", Bundle())


                        result_text.text = jsonObject.getString("message")

                        val fadeOut = AlphaAnimation(1f, 0f)
                        fadeOut.interpolator = AccelerateInterpolator() //and this
                        fadeOut.duration = 300
                        fadeOut.setAnimationListener(object : Animation.AnimationListener {
                            override fun onAnimationStart(animation: Animation) {

                            }

                            override fun onAnimationEnd(animation: Animation) {
                                email_view.visibility = View.INVISIBLE
                            }

                            override fun onAnimationRepeat(animation: Animation) {

                            }
                        })


                        val fadeIn = AlphaAnimation(0f, 1f)
                        fadeIn.interpolator = DecelerateInterpolator() //projects_add this
                        fadeIn.duration = 300
                        fadeIn.startOffset = 300
                        fadeIn.fillAfter = true


                        result_text.visibility = View.VISIBLE

                        email_view.startAnimation(fadeOut)
                        result_view.startAnimation(fadeIn)


                    } else {

                        FirebaseAnalytics.getInstance(applicationContext).logEvent("login_forgot_failed", Bundle())


                        error_text.text = jsonObject.getString("message")

                        val fadeOut = AlphaAnimation(1f, 0f)
                        fadeOut.interpolator = AccelerateInterpolator()
                        fadeOut.duration = 300
                        fadeOut.setAnimationListener(object : Animation.AnimationListener {
                            override fun onAnimationStart(animation: Animation) {}

                            override fun onAnimationEnd(animation: Animation) {
                                email_view.visibility = View.GONE
                                result_view.visibility = View.GONE
                            }

                            override fun onAnimationRepeat(animation: Animation) {}
                        })


                        email_view.startAnimation(fadeOut)


                        val fadeIn = AlphaAnimation(0f, 1f)
                        fadeIn.interpolator = DecelerateInterpolator() //projects_add this
                        fadeIn.duration = 300
                        fadeIn.startOffset = 300
                        fadeIn.fillAfter = true

                        error_view.visibility = View.VISIBLE
                        error_view.startAnimation(fadeIn)
                    }

                } catch (e: JSONException) {
                    e.printStackTrace()
                    FirebaseCrash.report(e)
                }

            }
        }
    }


    @Suppress("UNUSED_PARAMETER")
    fun reset(view: View) {

        val fadeIn = AlphaAnimation(0f, 1f)
        fadeIn.interpolator = DecelerateInterpolator() //projects_add this
        fadeIn.duration = 300
        fadeIn.startOffset = 300
        fadeIn.fillAfter = true

        val fadeOut = AlphaAnimation(1f, 0f)
        fadeOut.interpolator = AccelerateInterpolator()
        fadeOut.duration = 300
        fadeOut.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {}

            override fun onAnimationEnd(animation: Animation) {
                email_view.visibility = View.VISIBLE
                pacman_progress.visibility = View.INVISIBLE
                email_view.startAnimation(fadeIn)
                error_view.visibility = View.GONE
            }

            override fun onAnimationRepeat(animation: Animation) {}
        })


        error_view.startAnimation(fadeOut)

        email.showKeyboard()


    }


    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle item selection
        when (item.itemId) {
            android.R.id.home -> {
                background.hideKeyboard()
                startActivity(Intent(this, LoginActivity::class.java))
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onBackPressed() {
        startActivity(Intent(this, LoginActivity::class.java))
    }

}
